File: cs/ParkingTest/ParkingLotTest.cs
using parking;
using System;
using Xunit;

namespace ParkingTest
{
    public class ParkingLotTest
    {
        private readonly ParkingLot _parkingLot;

        public ParkingLotTest()
        {
            _parkingLot = new ParkingLot(2);
        }

        [Fact]
        public void Should_return_a_ticket_when_given_a_car()
        {
            var car = new Car("川A 1B2C3");
            var ticket = _parkingLot.Park(car);
            Assert.NotNull(ticket);
        }

        [Fact]
        public void Should_return_a_car_when_given_a_ticket()
        {
            var car = new Car("川A 1B2C3");
            var ticket = _parkingLot.Park(car);

            var takeCar = _parkingLot.Take(ticket);

            Assert.Equal(car, takeCar);
        }

        [Fact]
        public void Should_throw_an_exception_when_given_too_many_cars()
        {
            var car1 = new Car("川A 11111");
            var car2 = new Car("川A 22222");
            var car3 = new Car("川A 33333");
            _parkingLot.Park(car1);
            _parkingLot.Park(car2);

            Assert.Throws<Exception>(() => _parkingLot.Park(car3));
        }

        [Fact]
        public void Should_throw_an_exception_when_given_an_invalid_ticket()
        {
            var car = new Car("川A 1B2C3");
            var ticket = _parkingLot.Park(car);
            _parkingLot.Take(ticket);

            Assert.Throws<Exception>(() => _parkingLot.Take(ticket));
        }
    }
}

File: kt/src/test/kotlin/com/github/scglwsj/parking/ParkingLotTest.kt
package com.github.scglwsj.parking

import com.github.scglwsj.parking.lot.Car
import com.github.scglwsj.parking.lot.ParkingLot
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows


internal class ParkingLotTest {
    private lateinit var parkingLot: ParkingLot

    @BeforeEach
    fun setUp() {
        parkingLot = ParkingLot(2)
    }

    @Test
    fun `should return ticket when given a car`() {
        val car = Car("川A 1B2C3")
        val ticket = parkingLot.park(car)
        assertNotNull(ticket)
    }

    @Test
    fun `should return a car when given a ticket`() {
        val car = Car("川A 1B2C3")
        val takeCar = parkingLot.park(car)
            .let { parkingLot.take(it) }
        assertEquals(car, takeCar)
    }

    @Test
    fun `should throw an exception when given too many cars`() {
        val car1 = Car("川A 11111")
        val car2 = Car("川A 22222")
        val car3 = Car("川A 33333")
        parkingLot.park(car1)
        parkingLot.park(car2)

        assertThrows<Exception> {
            parkingLot.park(car3)
        }
    }

    @Test
    fun `should throw an exception when given an invalid ticket`() {
        val car = Car("川A 1B2C3")
        val ticket = parkingLot.park(car)
        parkingLot.take(ticket)

        assertThrows<Exception> {
            parkingLot.take(ticket)
        }
    }
}
